nomad/structs/structs.go

```diff
@@ -3449,6 +3463,112 @@ func (v *Vault) Validate() error {
 }
 
 const (
+	// DeploymentStatuses are the various states a deployment can be be in
+	DeploymentStatusRunning    = "running"
+	DeploymentStatusFailed     = "failed"
+	DeploymentStatusSuccessful = "successful"
+	DeploymentStatusCancelled  = "cancelled"
+	DeploymentStatusPaused     = "paused"
+)
+
+// Deployment is the object that represents a job deployment which is used to
+// transistion a job between versions.
+type Deployment struct {
+	// ID is a generated UUID for the deployment
+	ID string
+
+	// JobID is the job the deployment is created for
+	JobID string
+
+	// JobVersion is the version of the job at which the deployment is tracking
+	JobVersion uint64
+
+	// JobModifyIndex is the modify index of the job at which the deployment is tracking
+	JobModifyIndex uint64
+
+	// JobCreateIndex is the create index of the job which the deployment is
+	// tracking. It is needed so that if the job gets stopped and reran we can
+	// present the correct list of deployments for the job and not old ones.
+	JobCreateIndex uint64
+
+	// TaskGroups is the set of task groups effected by the deployment and their
+	// current deployment status.
+	TaskGroups map[string]*DeploymentState
+
+	// The status of the deployment
+	Status string
+
+	// StatusDescription allows a human readable description of the deployment
+	// status.
+	StatusDescription string
+
+	CreateIndex uint64
+	ModifyIndex uint64
+}
+
+func (d *Deployment) Copy() *Deployment {
+	c := &Deployment{}
+	*c = *d
+
+	c.TaskGroups = nil
+	if l := len(d.TaskGroups); l != 0 {
```

Comment:
I wouldn't guard on this, otherwise you could end up having two Deployments point to the same zero sized map, and then cause data sharing